sys/mips/cheri/locore.S

```diff
@@ -0,0 +1,330 @@
+	REG_LI		t0, CHERI_CAP_KERN_BASE
+	csetoffset	$c27, $c27, t0
+	REG_LI		t0, CHERI_CAP_KERN_LENGTH
+	csetbounds	$c27, $c27, t0
+	REG_LI		t0, CHERI_PERMS_KERNEL_CODE
+	candperm	$c27, $c27, t0
+
+	/* Preserve a copy in KCC for exception handlers.  */
+
+	csetkcc		$c27
+
+	/* Install the new PCC. */
+	REG_LI		t0, CHERI_CAP_KERN_BASE
+	cgetpcc		CHERI_REG_C29
+	cgetoffset	t1, CHERI_REG_C29			# 1
+	PTR_SUBU	t1, t1, t0				# 2
+	PTR_ADDIU	t1, t1, (4 * 7)				# 3
+	csetoffset	$c27, $c27, t1	# 4
+	cjr		$c27				# 5
+	nop							# 6
+	# 7 (land here)
+
+	/*
+	 * Initialize stack and call machine startup.
+	 */
+	cmove		CHERI_REG_STC, $c28
+	ABSRELOC_LA(t0, _C_LABEL(pcpu_space))
+	csetoffset	CHERI_REG_STC, CHERI_REG_STC, t0
+	REG_LI		t0, PCPU_SIZE
+	cincoffset	CHERI_REG_STC, CHERI_REG_STC, t0
+	REG_LI		t0, (PAGE_SIZE * 2) - PCPU_SIZE
+	csetbounds	CHERI_REG_STC, CHERI_REG_STC, t0	# Boot stack capability
+	PTR_SUBU	t0, t0, CALLFRAME_SIZ
+	cincoffset	CHERI_REG_STC, CHERI_REG_STC, t0	# Make room for initial callframe
+
+	/* XXX-AM: zero out ra so we catch anything that returns improperly */
+	move	ra, zero
+
+	/* Need to preserve a0-a3 for platform start */
+	move s0, a0
+	move s1, a1
+	move s2, a2
+	move s3, a3
+
+	/*
+	 * Process capability relocations.
+	 * This must be done before any PIC global access that uses the cap-table.
+	 * Pointers to the start and end of the __cap_relocs section are created
+	 * here in position-dependant code to avoid accessing the capability table.
+	 *
+	 * process_kernel_cap_relocs(reloc_start, reloc_end,
+	 *    code_cap, data_cap)
+	 */
+	ABSRELOC_LA(t0, _C_LABEL(__start___cap_relocs))
+	cmove		$c3, $c28
+	csetoffset	$c3, $c3, t0
+	ABSRELOC_LA(t1, _C_LABEL(__stop___cap_relocs))
+	PTR_SUBU	t0, t1, t0
+	csetbounds	$c3, $c3, t0
+	cincoffset	$c4, $c3, t0
+	cgetpccsetoffset	$c5, zero
+	cmove		$c6, $c28
+	ABSRELOC_LA(t0, _C_LABEL(process_kernel_cap_relocs))
+	cgetpcc		$c12
+	csetaddr	$c12, $c12, t0
+	cjalr	$c12, $c17
+	nop
+
+	/*
+	 * Process dynamic kernel relocations.
+	 *
+	 * process_kernel_dyn_relocs(code_cap, data_cap)
+	 */
+	cgetpccsetoffset	$c3, zero
+	cmove		$c4, $c28
+	ABSRELOC_LA(t0, _C_LABEL(process_kernel_dyn_relocs))
+	cgetpcc		$c12
+	csetaddr	$c12, $c12, t0
+	cjalr	$c12, $c17
+	nop
+
+	/* Now we can use cap-table relocations safely */
+	cmove	$c3, $c28
+	CAPCALL_LOAD($c12, _C_LABEL(cheri_init_capabilities))
+	cjalr	$c12, $c17
+	nop
+
+	/* Clear all remaining capabilities */
+	cclearlo	0xf7fe	# keep c0, c11 /* XXX-AM c0 will eventually be cleared */
+	cclearhi	0x9bff	# keep c26, c29, c30
+
+	/* Call the platform-specific startup code. */
+	move a0, s0
+	move a1, s1
+	move a2, s2
+	move a3, s3
+	CAPCALL_LOAD($c12, _C_LABEL(platform_start))
+	cjalr	$c12, $c17
+	nop
+
+	/*
+	 * thread0 set thread0 kstack to the bottom of the stack
+	 * page, realign thread0 kstack and start using it.
+	 */
+	CAPTABLE_LOAD($c3, _C_LABEL(thread0_st))
+	clc	CHERI_REG_STC, zero, TD_KSTACK($c3)
+	cgetlen	a0, CHERI_REG_STC
+	REG_LI	t0, ~7
```

Comment:
Alignment should probably not be required, I think it was introduced in the original MIPS locore because the PCB lives below the bottom of the stack and so it might not have the correct alignment, but it still seems redundant. We still have the PCB in the purecap kernel there, but it should now be moved in its own allocation to avoid representability issues. I'll do that in the purecap kernel and feed it back here.